crates/re_space_view_spatial/src/scene/parts/boxes2d.rs

```diff
@@ -0,0 +1,154 @@
+                    .annotation_info();
+                let color =
+                    annotation_info.color(color.map(move |c| c.to_array()).as_ref(), default_color);
+                let radius = radius.map_or(Size::AUTO, |r| Size::new_scene(r.0));
+                let label = annotation_info.label(label.map(|l| l.0).as_ref());
+
+                self.0.extend_bounding_box(
+                    macaw::BoundingBox {
+                        min: glam::Vec2::from(rect.top_left_corner()).extend(0.0),
+                        max: (glam::Vec2::from(rect.top_left_corner())
+                            + glam::vec2(rect.width(), rect.height()))
+                        .extend(0.0),
+                    },
+                    ent_context.world_from_obj,
+                );
+
+                let rectangle = line_batch
+                    .add_rectangle_outline_2d(
+                        rect.top_left_corner().into(),
+                        glam::vec2(rect.width(), 0.0),
+                        glam::vec2(0.0, rect.height()),
+                    )
+                    .color(color)
+                    .radius(radius)
+                    .picking_instance_id(instance_key_to_picking_id(
+                        instance_key,
+                        ent_view.num_instances(),
+                        ent_context.highlight.any_selection_highlight,
+                    ));
+
+                if let Some(outline_mask_ids) = ent_context
+                    .highlight
+                    .instances
+                    .get(&instance_hash.instance_key)
+                {
+                    rectangle.outline_mask_ids(*outline_mask_ids);
+                }
+
+                if let Some(label) = label {
+                    self.0.ui_labels.push(UiLabel {
+                        text: label,
+                        color,
+                        target: UiLabelTarget::Rect(egui::Rect::from_min_size(
+                            rect.top_left_corner().into(),
+                            egui::vec2(rect.width(), rect.height()),
+                        )),
+                        labeled_instance: instance_hash,
+                    });
+                }
+            },
+        )
+    }
+}
+
+impl ScenePart<SpatialSpaceViewClass> for Boxes2DPart {
+    fn archetype(&self) -> ArchetypeDefinition {
+        vec1::vec1![
+            Rect2D::name(),
+            InstanceKey::name(),
+            ColorRGBA::name(),
+            Radius::name(),
+            Label::name(),
+            ClassId::name(),
+        ]
+    }
+
+    fn populate(
+        &mut self,
+        ctx: &mut ViewerContext<'_>,
+        query: &SceneQuery<'_>,
+        _space_view_state: &SpatialSpaceViewState,
+        scene_context: &SpatialSceneContext,
+        highlights: &SpaceViewHighlights,
+    ) -> Vec<re_renderer::QueueableDrawData> {
+        re_tracing::profile_scope!("Boxes2DPart");
+
+        process_entity_views::<Rect2D, 6, _>(
```

Comment:
This is very error prone. If I change this `6` to a `5` it all compiles.

What is even this `6`? It is named `N`, which is the opposite of descriptive.

Is it the number of components in the archetype? Then `process_entity_views` could have a runtime check for that at least

Reply:
It is quite unfortunate that we sometimes treat the number of components as being dynamic, and sometimes as being static.

It seems to me we could easily re-write `query_primary_with_history` and `range_entity_with_primary` to take the number components as a runtime argument instead of a compile-time one. That would remove this big footgun, as `process_entity_views` would simply pass the archetype on to those functions. That would also reduce the amount of monomorphization, potentially improving compile times and reducing binary size.

Reply:
Yeah, I also really dislike this one. I wanted this it to be defined statically but didn't manage to do so so far. Maybe now that I shifted known types "further up" in the framework maybe it works 🤔 
I think it's not actually necessary to be statically passed around but for that I would have needed to touch stuff in the query and I really didn't want to go there on this monster PR...

Reply:
As discussed in stand-up, this should effectively go away with future code-gen.

`Rect2D` becomes an Archetype which knows it has 6 components at compile time.  Whether we use that information to generate static or dynamic code actually becomes a code-gen implementation question.